File: C++/Tarjan's_algorithm.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <algorithm>

using namespace std;

class Graph {
    int V; // Number of vertices

    // Adjacency list representation of graph
    vector<vector<int>> adj;

    // Helper function for Tarjan's algorithm
    void tarjanSCC(int u, vector<int>& disc, vector<int>& low, stack<int>& st, vector<bool>& stackMember, vector<vector<int>>& SCCs, int& time) {
        disc[u] = low[u] = ++time;
        st.push(u);
        stackMember[u] = true;

        for (int v : adj[u]) {
            if (disc[v] == -1) { // If v is not visited yet
                tarjanSCC(v, disc, low, st, stackMember, SCCs, time);
                low[u] = min(low[u], low[v]);
            } else if (stackMember[v]) { // If v is in the stack
                low[u] = min(low[u], disc[v]);
            }
        }

        // If u is the root of an SCC
        if (low[u] == disc[u]) {
            vector<int> currSCC;
            while (true) {
                int v = st.top();
                st.pop();
                stackMember[v] = false;
                currSCC.push_back(v);
                if (v == u)
                    break;
            }
            SCCs.push_back(currSCC);
        }
    }

public:
    Graph(int V) : V(V) {
        adj.resize(V);
    }

    // Function to add an edge to the graph
    void addEdge(int u, int v) {
        adj[u].push_back(v);
    }

    // Function to find strongly connected components using Tarjan's algorithm
    vector<vector<int>> findSCCs() {
        vector<int> disc(V, -1);
        vector<int> low(V, -1);
        vector<bool> stackMember(V, false);
        stack<int> st;
        vector<vector<int>> SCCs;
        int time = 0;

        for (int i = 0; i < V; ++i) {
            if (disc[i] == -1) {
                tarjanSCC(i, disc, low, st, stackMember, SCCs, time);
            }
        }

        return SCCs;
    }
};

int main() {
    // Create a graph given in the example
    Graph g(5);
    g.addEdge(0, 1);
    g.addEdge(1, 2);
    g.addEdge(2, 0);
    g.addEdge(1, 3);
    g.addEdge(3, 4);

    // Find strongly connected components
    vector<vector<int>> SCCs = g.findSCCs();

    // Output SCCs
    cout << "Strongly Connected Components:\n";
    for (const auto& SCC : SCCs) {
        for (int v : SCC) {
            cout << v << " ";
        }
        cout << "\n";
    }

    return 0;
}

File: Java/Tarjan's_algorithm.java
import java.util.*;

class TarjanSCC {
    private int V; // Number of vertices
    private List<Integer>[] adjList; // Adjacency list representation of the graph
    private int[] disc; // Discovery time of vertices
    private int[] low; // Earliest visited vertex reachable from subtree rooted with current vertex
    private boolean[] onStack; // To check if a vertex is on the recursion stack
    private Stack<Integer> stack; // Stack to store vertices in current SCC
    private List<List<Integer>> sccList; // List to store all SCCs

    public TarjanSCC(int vertices) {
        this.V = vertices;
        this.adjList = new ArrayList[V];
        for (int i = 0; i < V; i++)
            adjList[i] = new ArrayList<>();
        this.disc = new int[V];
        this.low = new int[V];
        this.onStack = new boolean[V];
        this.stack = new Stack<>();
        this.sccList = new ArrayList<>();
    }

    // Function to add an edge to the graph
    public void addEdge(int u, int v) {
        adjList[u].add(v);
    }

    // Function to perform Tarjan's Algorithm to find SCCs
    public void tarjanSCC() {
        // Initialize discovery and low values
        Arrays.fill(disc, -1);
        Arrays.fill(low, -1);

        for (int i = 0; i < V; i++) {
            if (disc[i] == -1)
                tarjanDFS(i);
        }
    }

    // DFS traversal for Tarjan's Algorithm
    private void tarjanDFS(int u) {
        disc[u] = low[u] = stack.size();
        stack.push(u);
        onStack[u] = true;

        for (int v : adjList[u]) {
            if (disc[v] == -1) {
                tarjanDFS(v);
                low[u] = Math.min(low[u], low[v]);
            } else if (onStack[v]) {
                low[u] = Math.min(low[u], disc[v]);
            }
        }

        if (low[u] == disc[u]) {
            List<Integer> component = new ArrayList<>();
            int w;
            do {
                w = stack.pop();
                onStack[w] = false;
                component.add(w);
            } while (w != u);
            sccList.add(component);
        }
    }

    // Function to print all SCCs
    public void printSCCs() {
        System.out.println("Strongly Connected Components:");
        for (List<Integer> component : sccList) {
            System.out.println(component);
        }
    }
}

public class Main {
    public static void main(String[] args) {
        TarjanSCC g = new TarjanSCC(5);
        g.addEdge(0, 1);
        g.addEdge(1, 2);
        g.addEdge(2, 0);
        g.addEdge(3, 0);
        g.addEdge(3, 4);

        g.tarjanSCC();
        g.printSCCs();
    }
}
